src/optimizer/query_rewrite.c

```diff
@@ -3499,6 +3523,870 @@ qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE
+ *         2. Not an inner join or natural join.
+ *         3. CTEs or derived tables.
+ *         4. No foreign key.
+ */
+static bool
+qo_check_foreign_keys_referencing_primary_key_in_child_spec (PARSER_CONTEXT * parser, PT_NODE * query,
+							     QO_REDUCE_REFERENCE_INFO * reduce_reference_info)
+{
+  PT_NODE *curr_fk_spec = NULL;
+  MOP curr_fk_mop = NULL;
+  SM_CLASS_CONSTRAINT *curr_fk_cons = NULL;
+  bool is_reducible = false;
+
+  assert (parser != NULL && query != NULL);
+  assert (reduce_reference_info != NULL);
+  assert (reduce_reference_info->pk_spec != NULL);
+  assert (reduce_reference_info->pk_mop != NULL);
+  assert (reduce_reference_info->pk_cons != NULL);
+  assert (reduce_reference_info->fk_spec != NULL);
+  assert (reduce_reference_info->reduce_pred_point_list != NULL);
+
+  reduce_reference_info->fk_cons == NULL;
+  if (reduce_reference_info->append_not_null_pred_list != NULL)
+    {
+      parser_free_tree (parser, reduce_reference_info->append_not_null_pred_list);
+      reduce_reference_info->append_not_null_pred_list = NULL;
+    }
+
+  curr_fk_spec = reduce_reference_info->fk_spec;
+  assert (PT_NODE_IS_SPEC (curr_fk_spec));
+
+  er_log_debug (ARG_FILE_LINE, "%s: check child. (spec: %s)\n", __func__, pt_print_alias (parser, curr_fk_spec));
+
+  /* PT_ALL is not supported. */
+  if (curr_fk_spec->info.spec.only_all == PT_ALL)
+    {
+      goto exit_on_fail_with_exclude;
+    }
+
+  switch (PT_SPEC_JOIN_TYPE (curr_fk_spec))
+    {
+    case PT_JOIN_NONE:
+      [[fallthrough]];
+    case PT_JOIN_INNER:
+      [[fallthrough]];
+    case PT_JOIN_NATURAL:
+      break;
+
+    case PT_JOIN_CROSS:
+      [[fallthrough]];
+    case PT_JOIN_LEFT_OUTER:
+      [[fallthrough]];
+    case PT_JOIN_RIGHT_OUTER:
+      [[fallthrough]];
+    case PT_JOIN_FULL_OUTER:
+      [[fallthrough]];
+    case PT_JOIN_UNION:
+      goto exit_on_fail_with_exclude;
+    }
+
+  PT_SPEC_GET_DB_OBJECT (curr_fk_spec, curr_fk_mop);
+  if (curr_fk_mop == NULL)
+    {
+      /* CTEs and derived tables are excluded. */
+      goto exit_on_fail_with_exclude;
+    }
+
+  for (curr_fk_cons = sm_class_constraints (curr_fk_mop); curr_fk_cons != NULL; curr_fk_cons = curr_fk_cons->next)
+    {
+      if (curr_fk_cons->type != SM_CONSTRAINT_FOREIGN_KEY)
+	{
+	  continue;
+	}
+
+      reduce_reference_info->fk_cons = curr_fk_cons;
+
+      if (qo_check_foreign_key_referencing_primary_key_in_child_spec (parser, query, reduce_reference_info))
+	{
+	  return true;
+	}
+      else
+	{
+	  if (er_has_error ())
+	    {
+	      goto exit_on_fail;
+	    }
+
+	  continue;
+	}
+    }
+
+  assert (curr_fk_cons == NULL);
+
+  if (reduce_reference_info->fk_cons == NULL)
+    {
+      /* No foreign key */
+      goto exit_on_fail_with_exclude;
+    }
+
+  goto exit_on_fail;
+
+exit_on_fail_with_exclude:
+  reduce_reference_info->exclude_fk_spec_point_list =
+    parser_append_node (pt_point (parser, curr_fk_spec), reduce_reference_info->exclude_fk_spec_point_list);
```

Comment:
exclude "return false" intentionally to fall through?

Reply:
It is intended.
I will add comments below.
```
/* fallthrough */
```